examples/structured_data/movielens_recommendations_transformers.py

```diff
@@ -0,0 +1,606 @@
+"""
+
+"""
+## Setup
+"""
+
+import os
+import math
+from zipfile import ZipFile
+from urllib.request import urlretrieve
+import pandas as pd
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+from tensorflow.keras.layers.experimental.preprocessing import StringLookup
+from sklearn.model_selection import train_test_split
+
+"""
+## Prepare the data
+
+### Download and prepare the DataFrames
+
+First, let's download the movielens data.
+
+The downloaded folder will contain three data files: `users.dat`, `movies.dat`,
+and `ratings.dat`.
+"""
+
+urlretrieve("http://files.grouplens.org/datasets/movielens/ml-1m.zip", "movielens.zip")
+zip_ref = ZipFile("movielens.zip", "r").extractall()
+
+"""
+Then, we load the data into pandas DataFrames with their proper column names.
+"""
+
+users = pd.read_csv(
+    "ml-1m/users.dat",
+    sep="::",
+    names=["user_id", "sex", "age_group", "occupation", "zip_code"],
+)
+
+ratings = pd.read_csv(
+    "ml-1m/ratings.dat",
+    sep="::",
+    names=["user_id", "movie_id", "rating", "unix_timestamp"],
+)
+
+movies = pd.read_csv(
+    "ml-1m/movies.dat", sep="::", names=["movie_id", "title", "genres"]
+)
+
+"""
+Here, we do some simple data processing to fix the data types of the columns.
+"""
+
+users["user_id"] = users["user_id"].apply(lambda x: f"user_{x}")
+users["age_group"] = users["age_group"].apply(lambda x: f"group_{x}")
+users["occupation"] = users["occupation"].apply(lambda x: f"occupation_{x}")
+
+movies["movie_id"] = movies["movie_id"].apply(lambda x: f"movie_{x}")
+
+ratings["movie_id"] = ratings["movie_id"].apply(lambda x: f"movie_{x}")
+ratings["user_id"] = ratings["user_id"].apply(lambda x: f"user_{x}")
+ratings["rating"] = ratings["rating"].apply(lambda x: float(x))
+
+"""
+Each movie has multiple genres. We split them into separate columns in the `movies`
+DataFrame.
+"""
+
+genres = [
+    "Action",
+    "Adventure",
+    "Animation",
+    "Children's",
+    "Comedy",
+    "Crime",
+    "Documentary",
+    "Drama",
+    "Fantasy",
+    "Film-Noir",
+    "Horror",
+    "Musical",
+    "Mystery",
+    "Romance",
+    "Sci-Fi",
+    "Thriller",
+    "War",
+    "Western",
+]
+
+for genre in genres:
+    movies[genre] = movies["genres"].apply(
+        lambda values: int(genre in values.split("|"))
+    )
+
+"""
+Now, let's display some records from each DataFrame.
+"""
+
+print(f"User count: {len(users.index)}")
+users.head()
+
+print(f"Movie count: {len(movies.index)}")
+movies.sample(5).T
+
+print(f"Rating count: {len(ratings.index)}")
+ratings.head()
+
+"""
+### Transform the movie ratings data into sequences
+
+First, let's sort the the ratings data using the `unix_timestamp`, and then group the
+`movie_id` values and the `rating` values by `user_id`.
+
+The output DataFrame will have a record for each `user_id`, with two ordered lists
+(sorted by rating datetime): the movies they have rated, and their ratings of these movies.
```

Comment:
Suggestion: "they" -> "users" for clarity

```suggestion
(sorted by rating datetime): the movies and users' ratings of these movies.
```